riscv_analysis/src/cfg/function.rs

```diff
@@ -9,12 +9,27 @@ use super::CfgNode;
 
 #[derive(Debug, PartialEq, Eq)]
 pub struct Function {
-    pub nodes: Vec<Rc<CfgNode>>,
-    pub entry: Rc<CfgNode>, // Is only a FuncEntry node
-    pub exit: Rc<CfgNode>,  // Multiple exit points will be converted to
-    // a single exit point
+    /// Labels for the entry point of this function
+    labels: HashSet<With<LabelString>>,
+
+    /// List of all nodes in the function. May not be in any particular order.
+    nodes: RefCell<Vec<Rc<CfgNode>>>,
+
+    /// Entry node of the function.
+    pub entry: Rc<CfgNode>,
+
+    /// Exit node of the function. Multiple exit points will be converted to a
+    /// single exit point.
+    exit: RefCell<Rc<CfgNode>>,
+
     /// The registers that are set ever in the function
-    pub defs: HashSet<Register>,
+    defs: RefCell<HashSet<Register>>,
+}
+
+impl Hash for Function {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.name().hash(state);
```

Comment:
We should add some `uuid` field to the function and use that as the `Hash`. It's possible to change the `HashSet` after the function is created, which would invalidate the hash and lead to undefined behaviour.